src/bigistore.c

```diff
@@ -647,6 +648,185 @@ Datum bigistore_each(PG_FUNCTION_ARGS)
+    if (rsinfo == NULL || !IsA(rsinfo, ReturnSetInfo))
+        ereport(ERROR,
+                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                 errmsg("set-valued function called in context that cannot accept a set")));
+    if (!(rsinfo->allowedModes & SFRM_Materialize))
+        ereport(ERROR,
+                (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                 errmsg("materialize mode required, but it is not allowed in this context")));
+
+    /* let the executor know we're sending back a tuplestore */
+    rsinfo->returnMode = SFRM_Materialize;
+
+    /* caller must fill these to return a non-empty result */
+    rsinfo->setResult = NULL;
+    rsinfo->setDesc = NULL;
+}
+
+
+/*
+ * return values per key for multiple bigistores as a set
+ */
+PG_FUNCTION_INFO_V1(join);
+Datum
+join(PG_FUNCTION_ARGS)
+{
+    MemoryContext    oldcontext;
+    /* needed for VARIADIC array deconstruction */
+    ArrayType       *input;
+    Datum           *i_data;
+    bool            *i_nulls;
+    int              arg_count;
+    int16            i_typlen;
+    bool             i_typbyval;
+    char             i_typalign;
+    Oid              i_eltype;
+    /* result tuple stuff */
+    TupleDesc        tupdesc;        /* the fcinfo tupdesc */
+    Tuplestorestate *tupstore;
+    bool            *nulls;
+    HeapTuple        tuple;
+    Datum           *dvalues;
+
+    int             *offset;        /* pair offset per arg */
+    int             *len;           /* is->len per arg */
+    BigIStorePair   **pairs;
+    ReturnSetInfo   *rsinfo;
+
+    prepTuplestoreResult(fcinfo);
+
+    /* get a tuple descriptor for our result type */
+    switch (get_call_result_type(fcinfo, NULL, &tupdesc))
+    {
+        case TYPEFUNC_COMPOSITE:
+            /* success */
+            break;
+        case TYPEFUNC_RECORD:
+            /* failed to determine actual type of RECORD */
+            ereport(ERROR,
+                    (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+                errmsg("function returning record called in context "
+                       "that cannot accept type record")));
+            break;
+        default:
+            /* result type isn't composite */
+            elog(ERROR, "return type must be a row type");
+            break;
+    }
+
+    rsinfo = (ReturnSetInfo *) fcinfo->resultinfo;
+    Assert(rsinfo->returnMode == SFRM_Materialize);
+
+    oldcontext          = MemoryContextSwitchTo(rsinfo->econtext->ecxt_per_query_memory);
+    tupstore            = tuplestore_begin_heap(true, false, work_mem);
+    rsinfo->setResult   = tupstore;
+    rsinfo->setDesc     = tupdesc;
+    MemoryContextSwitchTo(oldcontext);
+
+    /* deconstruct argument array */
+    input       = PG_GETARG_ARRAYTYPE_P(0);
+    i_eltype    = ARR_ELEMTYPE(input);
+    get_typlenbyvalalign(i_eltype, &i_typlen, &i_typbyval, &i_typalign);
+    deconstruct_array(input, i_eltype, i_typlen, i_typbyval, i_typalign,
```

Comment:
We should pass NULL instead of i_nulls; the functions is declared variadic, but the elements of the input array may still be null, which we likely need to prevent.